[FAB-17960] Ch.Part.API: Implement filerepo

caod123 opened the pull request:
#### Type of change

- New feature

#### Description

Introduce a general purpose file repo for persisting join block data
as well as supporting CFT on join and remove operations from the
registrar and file ledger.

#### Related issues

Task: [FAB-17960](https://jira.hyperledger.org/browse/FAB-17960)
Story: [FAB-15711](https://jira.hyperledger.org/browse/FAB-15711)
Epic: [FAB-17712](https://jira.hyperledger.org/browse/FAB-17712)

ghaskins commented:
Thank you @muralisrini 

LGTM

Reviewed-by: Greg Haskins <gregory.haskins@gmail.com>

caod123 committed:
[FAB-17960] Ch.Part.API: Implement filerepo

* Introduce a general purpose file repo for persisting join block data
as well as supporting CFT on join and remove operations from the
registrar and file ledger.

Signed-off-by: Danny Cao <dcao@us.ibm.com>

yacovm commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,95 @@
+// destination indicated by the FileSuffix.
+func (r *Repo) Save(channelName string, content []byte) error {
+	fileRepoDir := r.Options.FileRepoDir
+	dest := filepath.Join(fileRepoDir, channelName+r.Options.FileSuffix)
```
is it possible to have the channel name a string that contains the OS file separator? If yes then maybe we need to ensure it doesn't contain such?

caod123 replied:
I remember there are some fairly strict channel name rules, but I don't remember what characters are allowed in a channel name. I would imagine there should be some validation at the API level since you'd need to pass the channel name in via the join endpoint anyways. I do agree though that there probably could be some additional checks around the channel name.

caod123 replied:
From https://hyperledger-fabric.readthedocs.io/en/release-2.0/build_network.html#create-join-channel: 
```
Channel names must be all lower case, less than 250 characters long and match the regular expression [a-z][a-z0-9.-]*
```

yacovm replied:
the code above which i missed, does just that.

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,102 @@
+		return err
+	}
+
+	return os.Rename(tmpFile.Name(), dest)
```
**Close and rename do not imply sync** -- a good deck that exactly talk about this scenario https://www.slideshare.net/nan1nan1/eat-my-data

As you might know, `Flush` flushes the data present in the write buffer only to the disk (no fsync).

tock-ibm replied:
You have to use `Sync()` on the file, I think, to write both data and metadata to disk.

cendhu replied:
Wait for this pkg to be merged -- https://github.com/hyperledger/fabric/pull/1457 It has the required utility function.

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,102 @@
+}
+
+// Scan parses the directory and produce a list of file names, filtered by suffix.
+func (r *Repo) Scan() ([]string, error) {
```
Don't we need synchronization between `Scan()`, `Read()`, and `Save()` using a lock? If not, can you clarify how these operations can interleave and still provide consistency?

cendhu replied:
Nit. `List()` may be more expressive.

caod123 replied:
I don't think these functions would ever be called concurrently as saving occurs at the rest API level on receiving a join channel request; scan and load are called in order on orderer startup/registrar initialization to trigger onboarding for each join block.

cendhu replied:
I see. I thought of this package as a generic file repo that has no assumptions about how the caller would use it. With your explanation, my assumption no longer holds and I don't know whether non-generic pkgs can reside in the common.

yacovm replied:
If the API doesn't state it is thread safe the assumption of the user should be that it's not

caod123 replied:
Hmm good point. I think we may need to take a step back and evaluate whether this should be designed around general purpose usage or specifically for channel participation right now as a lot of the discussions below seem to hinge on this nuance. There's always the option of keeping it as specific to channel participation for now and in the future if we ever need to make it more general purpose, some small changes here and there could be included at that time. It's easy to say this could be used for other purposes, but right now I'm blanking on thinking of exactly what else we could use this for.

cendhu replied:
Adding a lock and making it thread-safe is very easy to do it for this pkg. Correct me If I am missing any challenges. That's why I suggested and I believe that making these kinds of packages generic would help in a bug-free code (in the future). Just take it as my philosophical view and I would better leave it to the orderer squad to choose what is best. At least in the ledger, it helps us and we don't need to check everytime whether the call we make is thread safe or not :-)

tock-ibm replied:
Thread safety is not just thread races - concurrent access - but also visibility. In the channel participation API, the `Scan` is used by the main go-routine, when initializing the server, `Save` by the http server request go-routine (which probably may spin multiple go-routines if there are multiple requests?), and the `Remove` by the internal go-routine servicing the follower.Chain, one for each chain. Since this is used by multiple threads, I think a thread safe implementation is needed.

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,102 @@
+	"github.com/hyperledger/fabric/common/configtx"
+)
+
+type Repo struct {
```
Don't we need a `Remove()` method on the Repo? I assume that it is missing because the consumer of this pkg does not need it.

caod123 replied:
Yes we've deferred Remove for now until it is necessary. Although given that once an orderer has fully onboarded, the join block is no longer needed and can be safely removed so I can probably include the `Remove` on this PR as well.

tock-ibm replied:
That is correct, we need `Remove` when on boarding is done. I think it should be added.

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,102 @@
+}
+
+type Options struct {
+	FileSuffix          string
```
I assume the file suffix as a type of entry stored?

caod123 replied:
the file suffix is the extension or suffix for the files we are saving in the file repo. For joinblocks this will likely be something like `.joinblock` so all joinblocks are saved as `<filerepo dir>/<channel>.joinblock`

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,102 @@
+
+type Options struct {
+	FileSuffix          string
+	TransientFilePrefix string
```
Isn't this field internal to this pkg? Should we need to export it?

caod123 replied:
There may be some use case for configuring the prefix, but currently it will probably always be `~` so we can probably remove it entirely from the options and leave it as an internal package constant.

tock-ibm replied:
I think - leave it exposed but apply a default `~` for when it is not set.

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,102 @@
+	defer func() {
+		if err != nil {
+			tmpFile.Close()
+			os.Remove(tmpFile.Name())
```
It doesn't guarantee a removal right? Given that there is an error already, I am not sure whether the removal/sync would even succeed.

caod123 replied:
This is best effort mainly. As Yoav mentioned, we can clear out temp files in another place as well.

cendhu commented on orderer/common/filerepo/filerepo_test.go:
```diff
@@ -0,0 +1,116 @@
+	}
+
+	t.Run("Successful load", func(t *testing.T) {
+		bytes, err := r.Load("testdata/mychannel.joinblock")
```
for a common `filerepo` pkg, do we really need two 17 KB files in the testdata? Is it not possible to use a simple file with a known data so that we can compare the content easily? Please correct me if I am missing something.

caod123 replied:
Probably not, I wanted to test with a proper block but because we don't care about block validation at the file repo level you're probably right and these can just be simple files with dummy data.

cendhu commented on orderer/common/filerepo/filerepo_test.go:
```diff
@@ -0,0 +1,116 @@
+		// Check file bytes
+		bytes, err := ioutil.ReadFile("testdata/newchannel.joinblock")
+		require.NoError(t, err)
+		assert.Equal(t, bytes, []byte("block-bytes"))
```
why are we using a mix of `require` and `assert`? can't we use `require` consistently?

caod123 replied:
I saw the slew of PRs recently for standardizing the ledger level tests to use `require` over `assert`. While inherently this doesn't introduce any problems, I feel mildly apprehensive that we may be overusing `require` when `assert` might be the proper usage. I think this may come down to what the intended intentions and proper usages of the `assert` and `require` packages were for initially. I'm fine changing all of these tests to using `require` exclusively though since it doesn't affect anything.

cendhu replied:
`require package is very similar to assert, except for it terminates test if assertion is not successful.` https://goinbigdata.com/testing-go-code-with-testify/ assert was introduced first and then require. 

@btl5037 also mentioned that a similar change happened at the SDK side a few days/weeks back. I do not have a strong opinion on the orderer pkg but my point was that this test uses a mix of both `require` and `assert` which look odd to me.

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,102 @@
+}
+
+// Load reads the file contents.
+func (r *Repo) Load(fileName string) ([]byte, error) {
```
while the `Save()` receives a channelName, shouldn't load also receive the channel name? Maybe I am asking this because I don't have clarity on how the consumer would use it.

caod123 replied:
I believe the idea for the consumer would be to scan/list the joinblocks in the file repo and then load as needed the filenames returned from the `Scan/List` function. I thought about this initially but after thinking about it more I'm not sure if we should prefer taking a channel name or file name inherently. The `filerepo` could ideally be used for more general purpose applications outside of channel participation related usages; however taking a file name in for the `Save` function would require changing some of the logic around the temp file path. Maybe for now for standardization purposes we can use the `filerepo` exclusively for channel participation and modify `Load` to receive the channel name as well. Alternatively we could leave it as is for now and refactor it on the fly as we come up with a better picture for how the consumer will use these functions.

cendhu commented on orderer/common/filerepo/filerepo_test.go:
```diff
@@ -0,0 +1,116 @@
+	}
+
+	channelName := r.FileToChannelName("testdata/mychannel.joinblock")
+	assert.Equal(t, channelName, "mychannel")
```
Shouldn't the order of argument be `expected` and then `actual` value?

caod123 replied:
I'm so used to gomega matchers which is the opposite 😅 having the actual come first.

cendhu commented on orderer/common/filerepo/filerepo_test.go:
```diff
@@ -0,0 +1,116 @@
+	t.Run("Successful scan", func(t *testing.T) {
+		files, err := r.Scan()
+		assert.NoError(t, err)
+		assert.Equal(t, files, []string{"mychannel.joinblock"})
```
Whenever testing a function with a loop, I would always prefer to have at least two iterations.

caod123 replied:
I believe it already iterates twice, once to add `mychannel.joinblock` and the other to filter out `mychannel.block` which doesn't match the file suffix pattern.

tock-ibm commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,102 @@
+// to a tmp file prefixed by the TransientFilePrefix and then moves the file to the final
+// destination indicated by the FileSuffix.
+func (r *Repo) Save(channelName string, content []byte) error {
+	err := configtx.ValidateChannelID(channelName)
```
I don't think we need to validate the channel name... channel names are very strict, and this would limit the generality of this package. W.r.t the channel participation API, we check the validity of the channel name when we first process the request, so no need to do it here.

caod123 replied:
I figured this was the case. Will remove thanks!

tock-ibm commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,102 @@
+	fileRepoDir := r.Options.FileRepoDir
+	dest := filepath.Join(fileRepoDir, channelName+r.Options.FileSuffix)
+
+	tmpFile, err := ioutil.TempFile(fileRepoDir, r.Options.TransientFilePrefix+channelName)
```
I have second thoughts on this... maybe it would be much cleaner to add the transient marker `~` as appended to the  suffix? like:
`mychannel.joinblock` vs. `mychannel.joinblock~` 
This seems to be closer to what many other programs do, and, this way there will never be a confusion with object names (i.e. channel names) starting with a `~`?

tock-ibm replied:
Plus, why a temp file? as we have the transient marker?

tock-ibm commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,102 @@
+	}
+
+	fileRepoDir := r.Options.FileRepoDir
+	dest := filepath.Join(fileRepoDir, channelName+r.Options.FileSuffix)
```
I think we want to fail early if `dest` above already exists, otherwise the os.Rename() will fail us later.

tock-ibm commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,102 @@
+	defer func() {
+		if err != nil {
+			tmpFile.Close()
+			os.Remove(tmpFile.Name())
```
This may leave transient files on the dir after crash. When we start using the repo again, after a crash, we should clean all the transient files. In the participation API we use `Scan` first, so that might be a good place to do it (once!). However, I think a better way is to (optionally) make `options` private, and build a constructor that cleans transient files. This way we force the user towards a safe and correct usage pattern.

```go
type Repo struct {
  options Options
}

func NewRepo( options Options) (*Repo, error) {
  //Create the directory if not there
  //Scan and remove transient files
  //Build a Repo and return it
}
```

tock-ibm commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,102 @@
+	"github.com/hyperledger/fabric/common/configtx"
+)
+
+type Repo struct {
```
I would consider being more specific about the file permissions:

```go
	repoFilePermPrivateRW os.FileMode = 0600
	repoDirPermPrivateRWX os.FileMode = 0700
```

In addition, consider creating a separate directory for every suffix, under the parent dir. This way the `*.joinblock` and `*.remove` files will be in different sub-directories. This would speed up the scan and avoid filesystem sync issues if we use two file-repo's in the same process with different suffixes. 
Another way to do this is to explicitly define a different file repo dir for different uses in the orderer.yaml, for example.

caod123 replied:
I was originally thinking of having separate repo dir's defined in the orderer.yaml, but your first suggestion produces less configuration which may be simpler.

tock-ibm commented on orderer/common/filerepo/filerepo_test.go:
```diff
@@ -0,0 +1,116 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestFileRepo_Save(t *testing.T) {
```
Add a test for saving and loading empty files with a `nil` or empty []byte content, for the `channel.remove` use case.

tock-ibm commented on orderer/common/filerepo/filerepo_test.go:
```diff
@@ -0,0 +1,116 @@
+	})
+}
+
+func TestFileRepo_Scan(t *testing.T) {
```
In the context of the above discussion about cleaning temp files after a crash, add a test case that puts some dummy temp files in the directory and test that the first `Scan` or `NewRepo` cleans them up (I prefer the latter).

caod123 commented:
Pushed an initial update addressing the comments from above. Will update once #1457 is merged to use the fileutil functions for some of the file system operations.

tock-ibm commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,198 @@
+		if isTransientFile {
+			os.Remove(filepath.Join(repo.fileRepoDir, f.Name()))
+		}
+	}
```
Either validate that the fileSuffix does not end with the transientFileMarker, or match against suffix+marker, rather than just the marker... otherwise this is going to remove legitimate files.

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,199 @@
+			return nil, err
+		}
+		if isTransientFile {
+			os.Remove(filepath.Join(repo.fileRepoDir, f.Name()))
```
We need to call `sync()` on the parent directory.

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,199 @@
+	return func(r *Repo) {
+		r.transientFileMarker = marker
+	}
+}
```
Function pointers are not good for readability IMO. It is good when we want to add hooks (heavily used in linux kernel). I yet to read the updated code again since I reviewed a few days back. Is there no way to remove this?

caod123 replied:
I could change the `New` function to just take an optional variatic `transientFileMarker` list instead and read the first element in that list if it exists to override the default `transientFileMarker`. I was mainly trying to avoid just accepting a `transientFileMarker` argument and ignoring it if it was an empty string since the default would probably be to not pass an override.

sykesm replied:
This is using the OptionFunc pattern for extensible construction. I'm not sure why that's considered difficult to read. It's a flexible, pragmatic, and relatively common way of handling object construction in a friendly, backwards compatible way.

https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

The same pattern can be used in place of the "builder" pattern to minimize error handling to one call while building up a bunch of changes.

cendhu replied:
Whenever there is a function pointer and it's usage, I always had to spend some extra time to understand the code. If there is a simple way like `NewServer(config)` or `NewServer(config1, confg2)`, I spend less time to read and understand. This is what I meant regarding the readability. If this is a common pattern and used widely, then I might probably improve my readability skill once I find and read many such patterns.

As I mentioned in my first set of review, I do not see a strong reason to configure the `transientFileMarker`. It is just an internal logic to the `filerepo`. Given that, I am not convinced that we need a function pointer or configurable option.

sykesm commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,199 @@
+
+type repoOption func(r *Repo)
+
+func WithTransientFileMarker(marker string) repoOption {
```
Why do we keep doing this? An exported function should accept and return exported types.

Please export `RepoOption`.

caod123 replied:
I had initially left it unexported to leave defining functions that returned repoOptions to only be definable within this package. I guess that might have been too restrictive or unnecessary.

sykesm replied:
May also want to consider returning an error from the function since the constructor returns an error.

sykesm replied:
> I had initially left it unexported to leave defining functions that returned repoOptions to only be definable within this package. I guess that might have been too restrictive or unnecessary.

It's unnecessary. The reason you use options is so the implementation can access the internal details of the thing that's being constructed. The implementation of the function acting on `*Repo` must live in this package to do anything meaningful.

If you don't export the type, it's an inconsistent. It's no different than having a public constructor that returns an unexported implementation.

sykesm commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,199 @@
+	fileSuffix          string
+	transientFileMarker string
+
+	mu sync.Mutex
```
nit: grouping the lock with the fields it protects is a useful convention. We generally expect to see the lock declared and the fields is protects just below it.

caod123 commented:
Updated to use `fileutil` package

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,194 @@
+	}
+
+	for _, opt := range opts {
+		if err := opt(repo); err != nil {
```
additional response to @sykesm question on why I find this hard to read. 

1. I did go to definition by keeping the cursor on `opt`. Nothing happened.
2. I needed to track to opts and reach `ReadOption`
3. Then, I find that ReadOption is a function pointer and it provides a list of functions which the consumer could have called to get a function implementation that sets a value for a variable. 
4. The returned function implementation is called here. 

I understand that this allow the caller to configure only the required variables (i.e., optional variables) but adds more time to the code readability (& multi-level functions). As I do not see any strong reason for the caller to set `transientFileMarker` (as it is internal to the `filerepo` pkg), I am not convinced that this added code complexity is required. The default `~` looks fine to me. If others have a strong opinion that the added code complexity is justified and `transientFileMarker` needs to be configured by the caller, I will step aside.

caod123 replied:
I'll remove the configurability for now as I don't see an immediate need to override the default value. We can always revisit this down the line if an override is needed.

sykesm replied:
In this instance, fine. In general, this is the wrong way.

When it comes to constructing objects, we have:

1. Zero values
2. Struct literals that require nothing more than wiring
3. Constructors that take no arguments
4. Constructors that take specific arguments
5. Constructors that can evolve

Option 1 is best. It's the "make the zero value useful" goal of of the standard library.

Option 2 is good when evolving from option 1. This allows us to create a default behavior with the zero value and add customizability after the fact by injecting something to override behavior. When we set a field, we have an explicit name at the call site so we know what we're providing. We see a lot of this in the http package.

Option 3 is okay but limiting. It means that we're constructing something and there's no way to control it. That implies we end up having to modify the state after construction to make things useful or build global state around it to influence it.

Option 4 is rigid and confusing. The rigidity comes from the fact that there's never a backward compatible way to add or remove constructor arguments without impacting callers. It's confusing when we get long lists of arguments that share the same type. In this particular case, it looks like we have `n` strings. So, at the call site, I need to figure out the order and what they mean.

Option 5 is good. In this model we can easy extend the functionality of constructors. Each option is named so when we make the call, we know what we're providing. The options allow better encapsulation of the object as even the names of the fields we use to hold dependencies is not exposed to the caller.

Given how prevalent the option func pattern is in libraries, we should all be able to recognize it for what it is and embrace the benefits it provides when it comes to evolving a code base.

cendhu replied:
Thanks, @caod123 for removing this configuration option.

@sykesm Thanks for the nice summary. Especially what you mentioned in option 4 and 5 are nice -- no need to break backward compatibility when we add new configuration option -- didn't strike me. While optionalal func pattern might add a bit of overhead while reading it for the first time (as in my case), we will adjust to it as we see more patterns. I hope you don't mind that I am copying your suggestions to the [google doc](https://docs.google.com/document/d/1mU4S4U1dDpBk9AA180RMRB7g3OH887nqbcEPGD3px3Q/edit?usp=sharing) that I maintain so that these suggestions/recommendations are not lost within the PR comments.

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,169 @@
+
+	fileRepoDir := filepath.Join(repoParentDir, fileSuffix)
+
+	// Make the file repo if it doesn't already exist
```
Nit. Would be good to have comments to explain a decision or reasoning or complex side-effects. Other types of comments are distraction while reading the code.

caod123 commented:
@cendhu could you do another pass of this and let me know if it looks good to you? There hasn't been much activity on this since last week.

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,169 @@
+	repoFilePermPrivateRW os.FileMode = 0600
+)
+
+type Repo struct {
```
Nit. Doc comment for this exported type is missing.

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,169 @@
+			return nil, err
+		}
+		if isTransientFile {
+			os.Remove(filepath.Join(fileRepoDir, f.Name()))
```
Better to check the err returned by `os.Remove()`

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+func validateFileSuffix(fileSuffix string) error {
+	if len(fileSuffix) <= 0 {
```
Nit. as it can never be < 0
```suggestion
	if len(fileSuffix) == 0 {
```

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+// Remove removes the file associated with baseName from the file system.
+func (r *Repo) Remove(baseName string) error {
```
Unit test is missing.

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,169 @@
+
+	// Remove existing transient files in the repo
+	for _, f := range files {
+		transientFilePattern := "*" + fileSuffix + "~"
```
Nit. Can be kept outside the loop instead of creating it newly for each file.

cendhu replied:
Nit. `~` -- till we reach the line 74, the reader cannot understand what this mean. One way to fix this is by using a const variable to keep the default transientFileMarker.

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,169 @@
+	defer r.mu.Unlock()
+
+	fileName := fmt.Sprintf("%s.%s", baseName, r.fileSuffix)
+	dest := filepath.Join(r.fileRepoDir, fileName)
```
Nit. Given that these two lines occurs in three methods on `Repo`, it would be good to move them to a separate method called `baseToFileName()` (which would do an opposite of the existing `FileToBaseName()`)

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,169 @@
+		return nil, err
+	}
+
+	fileRepoDir := filepath.Join(repoParentDir, fileSuffix)
```
Nit. Using `fileSuffix` both as a sub-directory name and also as a suffix to the file name within the sub directory kind of confuses me while reading the code. I think the fileSuffix denotes the type of content being stored. One such type is the join block. We then create a sub-folder for the type and also add the type as a suffix to each file stored in the sub-folder.

Anyway, I do not have an alternative to the current folder structure that is shown below. Maybe the joinblock suffix looks redundant. 
```
/repo/joinblock
/repo/joinblock/channel1.joinblock
/repo/joinblock/channel2.joinblock
/repo/joinblock/channel3.joinblock~
```

caod123 replied:
Having separate subfolders helps to isolate different file repos from operating on the same folder. The suffix is still useful for filtering in case other files end up in the folder through other means.

cendhu replied:
> Having separate subfolders helps to isolate different file repos from operating on the same folder. The suffix is still useful for filtering in case other files end up in the folder through other means.

What other means? You mean admin creates some files manually?

caod123 replied:
Sure, I don't see it really happening but in the off chance the admin may backup a joinblock to the same directory or for some other reason accidentally move a separate file to the folder. In the end it doesn't really hurt to have the suffix in addition to the subfolder even if it introduces some minor redundancy. As for the subfolder existence itself, Yoav pointed out in a previous comment that having separate subfolders would speed up the scan and avoid filesystem sync issues if we use two file-repo's in the same process with different suffixes

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,169 @@
+	}
+
+	if strings.Contains(fileSuffix, string(os.PathSeparator)) {
+		return errors.Errorf("fileSuffix [%s] illegal, cannot contain os path separator", fileSuffix)
```
Nit.`fmt.Errorf`and `errors` pkg instead of `github.com/pkg/errors`

cendhu commented on orderer/common/filerepo/filerepo_test.go:
```diff
@@ -0,0 +1,168 @@
+
+	// Check tempfile was cleared out upon creating a new file repo
+	_, err = os.Stat(tempFilePath)
+	require.True(t, os.IsNotExist(err))
```
Nit. `require.NoFileExists` can be used here.

cendhu commented on orderer/common/filerepo/filerepo_test.go:
```diff
@@ -0,0 +1,168 @@
+
+			// Check tempfile doesn't exist
+			_, err = os.Stat(filePath + "~")
+			require.True(t, os.IsNotExist(err))
```
Nit. `require.NoFileExists` can be used here.

cendhu commented on orderer/common/filerepo/filerepo.go:
```diff
@@ -0,0 +1,169 @@
+	}
+
+	for _, f := range files {
+		isFileSuffix, err := filepath.Match("*"+r.fileSuffix, f.Name())
+		if err != nil {
+			return nil, err
+		}
+		if isFileSuffix {
+			repoFiles = append(repoFiles, f.Name())
+		}
```
Is this necessary? Given that methods on `Repo` are thread-safe and `New()` removes all transientFiles, do we need to perform the match based on fileSuffix. If it is necessary, please update the test to reflect this. Otherwise, remove this part of the code.

caod123 replied:
It's still useful to ensure other files are filtered out if they somehow end up there. The tests already reflect this by filtering out `testdata/joinblock/mychannel.block` and only returning `testdata/joinblock/mychannel.joinblock`

cendhu replied:
Probably, I missed that. If test is there, then it is fine. 

> if they somehow end up there

I do not understand. Are you referring to some attack scenario?

cendhu reviewed (commented):
Danny, Most of my comments are nit and hence, I leave these to you to decide. 

Once you add the missing test, I think the PR can be merged.

cendhu reviewed (approved):
Thanks, Danny. I am still not sure about the following two points 

1. having a fileSuffix as both folder name and the file suffix
2. having a filter logic in the List()

If required, it can be fixed in a separate PR after the discussion and need not to block this PR. 

In future, please try to not rebase in the middle of reviews. Otherwise, it is quite hard to find out what changed compared to the initial commits.
